### PR TITLE
[fix][txn] Fix transaction is not aborted when send or ACK failed

### PR DESCRIPTION
**Pull request by BewareMyPower**
### Motivation

In `TransactionImpl`, when `hasOpsFailed` is false, i.e. there is a send or ACK operation that failed, `commit()` will fail and `abort()` will be called. However, since the state is `COMMITTING`, the abort process will not be executed.

### Modifications
- Add the `checkState` method to remove duplicated code in `checkIfOpenOrCommitting` and `checkIfOpenOrAborting`.
- In the callback of `commit()`, check the state is `COMMITTING` instead of the same state check in `abort()`.
- Add `TransactionProduceTest#testCommitFailure` to verify when a send operation failed, the `commit()` method will change the state of the `Transaction` to `ABORTED` and the `TxnMeta` in TC will be removed.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/BewareMyPower/pulsar/pull/26

### CODE REVIEW

**Review by nicoloboschi — APPROVED** [review]
LGTM

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionProduceTest.java [hunk 1]
```diff
@@ -369,5 +373,24 @@ private int getPendingAckCount(String topic, String subscriptionName) throws Exc
...
+    @Test
+    public void testCommitFailure() throws Exception {
+        Transaction txn = pulsarClient.newTransaction().build().get();
+        final String topic = NAMESPACE1 + "/test-commit-failure";
+        @Cleanup
+        final Producer<byte[]> producer = pulsarClient.newProducer().topic(topic).create();
+        producer.newMessage(txn).value(new byte[1024 * 1024 * 10]).sendAsync();
+        try {
+            txn.commit().get();
+        } catch (ExecutionException e) {
```
Comment (congbobo184) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionProduceTest.java:385]:
```suggestion
        try {
            txn.commit().get();
            fail();
        } catch (ExecutionException e) {
```

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionProduceTest.java [hunk 2]
```diff
@@ -369,5 +373,24 @@ private int getPendingAckCount(String topic, String subscriptionName) throws Exc
...
+            Assert.assertTrue(e.getCause() instanceof PulsarClientException.TransactionHasOperationFailedException);
+            Assert.assertEquals(txn.getState(), Transaction.State.ABORTED);
+        }
+        try {
+            getPulsarServiceList().get(0).getTransactionMetadataStoreService().getTxnMeta(txn.getTxnID())
+                    .getNow(null);
+        } catch (CompletionException e) {
```
Comment (congbobo184) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionProduceTest.java:392]:
```suggestion
            getPulsarServiceList().get(0).getTransactionMetadataStoreService().getTxnMeta(txn.getTxnID())
                    .getNow(null);
                    fail();
        } catch (CompletionException e) {
```

**Review by congbobo184 — COMMENTED** [review]
Nice catch!

**Comment by BewareMyPower** [comment]
@congbobo184 Good suggestions. They are addressed now.

**Review by congbobo184 — APPROVED** [review]

**Comment by michaeljmarshall** [comment]
As discussed on the mailing list https://lists.apache.org/thread/w4jzk27qhtosgsz7l9bmhf1t7o9mxjhp, there is no plan to release 2.9.6, so I am going to remove the release/2.9.6 label

**Pull request merged**